### ISSUE TITLE
Support WPA3 on Wifi

### ISSUE THREAD

**Issue by schildbach** [Aircoookie/WLED#3040]
**Is your feature request related to a problem? Please describe.**
WLED 0.13.x on an ESP32 (Athom Addressable Strip Controller) connects to Wifi only via WPA2. With WPA2, several kind of attacks are possible that are not possible with WPA3.

**Describe the solution you'd like**
Support WPA3. The ESP32 SDK seems to be able to do this out of the box. Tasmota has enabled this too.

**Describe alternatives you've considered**
If WPA3 cannot be considered, supporting PMF (protected management frames) on WPA2 would provide a similar level of security.

**Comment by blazoncek** [comment]
Can you point to the documentation on this?
Especially for Arduino platform.

**Comment by schildbach** [comment]
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi-security.html

**Comment by schildbach** [comment]
https://www.arduino.cc/reference/en/libraries/wifi/ only mentions WPA2. I hope this doesn't mean Arduino doesn't support WPA3.

**Comment by schildbach** [comment]
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi-security.html#wpa3-personal

**Comment by schildbach** [comment]
https://www.arduino.cc/reference/en/libraries/wifi/ only mentions WPA2 and below. I hope this doesn't mean Arduino doesn't support WPA3.

**Comment by softhack007** [comment]
> https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi-security.html

The first mention of WPA3 security mode is in the IDF-v4.3.x docs. So it might be that WPA3 is only supported when building with the newer platform framework - WLED is still on ESP-IDF 3.3.6 (`platform = espressif32@3.5.0`) while you might need `platform = espressif32@ ^5.0.0` or later to have the WPA3 functionality included in the framework.

Currently we cannot simply "upgrade" the platform, due to crazy bugs and stability problems in ESP-IDF 4.x.y - WLED on esp32 would basicially become unusable and crash randomly. But OK, at least we could have WPA3 ;-) When espressif would finally get their act together and start testing/fixing regressions on esp32, then -maybe- WPA3 could come in future WLED release ...

**Comment by Jason2866** [comment]
@softhack007 Latest Arduino core 2.0.6 is stable. Tasmota does run rock solid with.
Code needs to be adopted to the changed APIs ;-)

**Comment by softhack007** [comment]
> @softhack007 Latest Arduino core 2.0.6 is stable.

As much as I want to belief you ... I've tried the latest core 2.0.6 on a -C3 and -S2, and what i saw was worse than 2.0.4/2.0.5. "Classic ESP32" also crashes too often, compared to the rock-solid build with platform version 3.5.0. Maybe we are not ready yet for new cores and newest ESP-IDF.

From my (limited) experiences, arduino-esp32 2.04/2.0.5  (and ESP-IDF 4.4.1 / 4.4.2) still seem to be the better choice when a new platform version is absolutely necessary.

**Comment by Jason2866** [comment]
Just can say Tasmota runs weeks without an reboot or WDT on ESP32, ESP32-S2, ESP32-S3 and the ESP32-C3. Yes, it needs changes in code, to get there.
Don't get me wrong, in no way i mean the next statement offending.
WLED is and was (until now) never rock solid on an ESP32.
Never managed to have an uptime over 2 months without an reboot or WDT.
With the old core I doubt you will manage this. The underlying IDF and Arduino code has too many bugs.
To be clear i love WLED and use it!!!

**Comment by blazoncek** [comment]
@Jason2866 could you help us with issues with LittleFS on regular ESP32 and Arduino core 3.5+?

**Comment by Jason2866** [comment]
The issue with LittleFS and the older cores is related to the not correct builded precompiled  bootloaders. There is no real fix for. Only moving to actual core (2.0.5 and 2.0.6)
Or you have to backport the way the bootloader is builded introduced with core 2.0.5 to older cores. One of more reasons why i started to maintain our own fork and compile the Arduino libs our self. 
In short flash mode is not always correctly initialized (from the Bootloader) and data gets corrupted when flash is accessed from LittleFS. That's not LittleFS fault. 
Probably you have a message like `corrupted dir pair`

**Comment by Jason2866** [comment]
So long story short. To avoid flash corruptions (not restricted to LittleFS!) cores before 2.0.5 can only be used in mode `DOUT`

**Comment by blazoncek** [comment]
That is an eye opener. Thank you.
We experimented with QIO and DIO and flash speed of 80MHz due to occasional flickering while frequently accessing flash and/or PSRAM during effect calculations.

To summarize:
- update to most recent bootloader
- use DOUT flash mode
- user more recent PIO platform if it is still compatible with WLED components
- try alternative flashing with esptool

Is that right?

**Comment by Jason2866** [comment]
Yes, right direction.

**Comment by MacMika** [comment]
same here. Are there any plans?

**Comment by Jason2866** [comment]
When WLED is moving to actual Arduino core 3.x. there will be WPA3 support.
As long this is not done there is no WPA3.

**Comment by schildbach** [comment]
It seems version 0.14.4 has WPA3, at least the version that on the Athom/IoTorero ESP32C3 controller, arriving by mail today.

If that's true, many thanks for implementing WPA3!

**Comment by schildbach** [comment]
Version 0.15.1.beta2 on Athom ESP32 "Music Addressable LED Strip Controller" still connects via WPA2.

I guess I will update the ESP32C3 to stock WLED too, to check if it retains its WPA3 capability.

**Comment by blazoncek** [comment]
WLED does not enforce or *enable* certain WiFi protocols, it uses whatever framework supports.
ESP32 uses ancient framework.

**Comment by schildbach** [comment]
So ESP32 uses a framework (Arduino core, I assume) version different from ESP32-C3? If so, where can I see the Arduino version used? Is that perhaps something that could be added to the "Info" screen?

**Comment by blazoncek** [comment]
It already is.

**Comment by martijnbuts** [comment]
I have wled version 15.3 with environment version v3.3.6. Is there any option to get WPA3 working?

**Comment by netmindz** [comment]
You don't say which board you are using @martijnbuts but if you look in the info page and see if the platform version is 3 or 4

If it's still 3, please try the esp32 V4 bin, that *might* have a support

**Comment by martijnbuts** [comment]
HHi Netmindz,

My apologies for the late response.

The board I’m using is an ESP32-WROOM-32. I’ve also tested the latest nightly build (v16.0.0) with the ESP32 environment version 4.4.8.240628. (is this the platform version you refer to? ) 

Using the same board with ESPHome (based on ESP-IDF), WPA3 works correctly. However, I haven’t yet tested WPA3 with ESPHome when using the Arduino framework.

WPA3 support is not strictly required for my setup—I could alternatively create a separate VLAN for WLED and move the ESP32 running WLED to that network segment. 

My main motivation for WPA3 is the support for management frame protection/encryption (PMF  https://www.wi-fi.org/beacon/philipp-ebbecke/protected-management-frames-enhance-wi-fi-network-security ) , which helps mitigate certain Wi-Fi attacks.

Thank you for your time and effort.

Kind regards,
Martijn

<img width="641" height="828" alt="Image" src="https://github.com/user-attachments/assets/9314b597-d35f-4fcb-a71b-a46b3424ddea" />

**Comment by netmindz** [comment]
@coderabbitai what version of the ESP IDF is required to support WPA3 ?

**Comment by netmindz** [comment]
Hi @martijnbuts 

It looks like we are running a new enough version of the platform such that you should already have WPA3 support 

Can you try running a WPA3 only network and see if you can connect?

**Comment by schildbach** [comment]
@netmindz My home router (a FritzBox, very common in Germany) doesn't have "WPA3 only". Rather, the highest setting is "WPA2 + WPA3".

Yet, WLED is the only firmware that still connects via WPA2 although it should be possible to use WPA3 on that hardware. Is it perhaps possible to restrict WLED to WPA3, or at least assign a higher priority to WPA3?

**Comment by Jason2866** [comment]
@schildbach There is no priority setting in IDF core. Would have to be implemented manually. Not recommended since this will add complexity and additional time for connect.
Normally this is done in the AP. Bad luck Fritz Box does not provide this feature. Probably done not to generate issues from users not knowing what the do.

Btw. removed WPA 3 support in our Tasmota Arduino framework based on IDF 5.x. It takes too much flash and RAM. Can be added when using HybridCompile mode.

**Comment by softhack007** [comment]
@Jason2866 would we need to add specific arduino-esp32 API calls for enabling WPA3 with the `WiFi` object? Similar to WPA2-enterprise which required some special handling in the WLED code?

**Comment by Jason2866** [comment]
@softhack007 No, it will connect to WPA3. The problem is the handshake between Fritz Box and the device. For some reason idk Fritz Box does not use WPA3 and uses WPA2. To solve such behaviour most routers have a setting to allow WPA3 **only**.

EDIT: Non consumer APs provide a lot of settings for WPA3. There is no WPA3 feature settings setup which is optimal for everything. After some trys i found a setup where the espressif devices always connects in mode WPA3 even there is WPA2 support enabled. And all other WiFi devices are still connecting in mode WPA3.
espressif wifi stack do not support all possible WPA3 ciphers. Not possible for a tiny MCU where wifi is "just" an add on feature and some ciphers needs a lot of resources.

**Comment by netmindz** [comment]
> Btw. removed WPA 3 support in our Tasmota Arduino framework based on IDF 5.x. It takes too much flash and RAM. Can be added when using HybridCompile mode.

Thanks very useful to know

**Comment by schildbach** [comment]
> [@softhack007](https://github.com/softhack007) No, it will connect to WPA3. The problem is the handshake between Fritz Box and the device. For some reason idk Fritz Box does not use WPA3 and uses WPA2. To solve such behaviour most routers have a setting to allow WPA3 **only**.

I don't think this issue is specific to FritzBox. None of Android, Windows, Linux, ESPHome and various propietary Google, Nest and Tado devices connect with WPA2, all use WPA3 as expected. And they all of course have the ability to fallback to WPA2 if necessary.

Rather, I think this is something in the WPA handling of WLED.

**Comment by softhack007** [comment]
Btw, are we repeating this discussion from tasmota? 

https://github.com/arendst/Tasmota/discussions/20472

**Comment by Jason2866** [comment]
Looks like ;-)

@schildbach WLED is WLED. Different design choices. Fork and change if you need different settings. It is Open Source.

**Comment by schildbach** [comment]
Well, Tasmota priorizes features over security. So I dropped it from my home, in most cases in favor of ESPHome.

WLED on the other hand so far hasn't made an "official" statement about this. To me it appears WPA3 is supported, but a bug prevents it from working in all cases.

**Comment by softhack007** [comment]
> Rather, I think this is something in the WPA handling of WLED.

@schildbach WLED does not really "handle" WPA, we just use the standard ``wifi`` feature of arduino-esp32. Basicially ``wifi.begin(...)`` and nothing more.

**Comment by schildbach** [comment]
> > Rather, I think this is something in the WPA handling of WLED.
> 
> [@schildbach](https://github.com/schildbach) WLED does not really "handle" WPA, we just use the standard `wifi` feature of arduino-esp32. Basicially `wifi.begin(...)` and nothing more.

Sure, I'm aware WLED isn't implementing the Wifi protocols themselves. When I say WLED, I mean the entire firmware, including all its linked dependencies.

**Comment by softhack007** [comment]
Ok, below info is directly from espressif docs for arduino-esp32 v2.0.17 (our latest 16.0 nightly uses v2.0.18 with the "tasmota" core)

Espressif states that "WPA3 support is resource-intensive", which could imply that 4MB flash becomes too small to store existing wled feature, plus WPA3 "ressource-intensive" extras.

@schildbach as 4MB esp32 are quite old hardware, would you say that dropping 4MB board support (=8MB flash is the new minimum) is an adequate price to pay for WPA3?


https://github.com/espressif/arduino-esp32/blob/5e19e086c43d0fa5e5a596497ff8f11a0a43f6c2/docs/source/troubleshooting.rst?plain=1#L107-L125

https://github.com/espressif/arduino-esp32/blob/5e19e086c43d0fa5e5a596497ff8f11a0a43f6c2/libraries/WiFi/src/WiFiSTA.cpp#L417-L425

**Comment by schildbach** [comment]
> as 4MB esp32 are quite old hardware, would you say that dropping 4MB board support (=8MB flash is the new minimum) is an adequate price to pay for WPA3?

Personally, I've got an Athom LS8P and an LS8P-dual, both of which I believe come equipped with 8 MB flash. However, I'd perhaps look at your device statistics to get a feel for the userbase. If people use WLED for smart bulbs, that's probably 4 MB territory.

In any case, enabling WPA3 for an 8 MB image would be a great start!

**Comment by Jason2866** [comment]
Quote myself. WPA3 support for 8MB devices can be done:

> Btw. removed WPA 3 support in our Tasmota Arduino framework based on IDF 5.x. It takes too much flash and RAM. Can be added when using HybridCompile mode.
